opendevin/controller/agent_controller.py

```diff
@@ -83,106 +76,79 @@ def __init__(
+    async def report_error(self, message: str):
+        await self.event_stream.add_event(ErrorObservation(message), EventSource.AGENT)
 
-    async def add_history(
-        self, action: Action, observation: Observation, add_to_stream=True
-    ):
-        if not isinstance(action, Action):
-            raise TypeError(
-                f'action must be an instance of Action, got {type(action).__name__} instead'
-            )
-        if not isinstance(observation, Observation):
-            raise TypeError(
-                f'observation must be an instance of Observation, got {type(observation).__name__} instead'
-            )
+    async def add_history(self, action: Action, observation: Observation):
         self.state.history.append((action, observation))
         self.state.updated_info.append((action, observation))
-        if add_to_stream:
-            await self.event_stream.add_event(action, EventSource.AGENT)
-            await self.event_stream.add_event(observation, EventSource.AGENT)
-
-    async def _run(self):
-        if self._agent_state != AgentState.RUNNING:
-            raise ValueError('Task is not in running state')
 
-        for i in range(self._cur_step, self.max_iterations):
-            self._cur_step = i
+    async def _start_step_loop(self):
+        while True:
             try:
-                finished = await self.step(i)
-                if finished:
-                    await self.set_agent_state_to(AgentState.FINISHED)
-                    break
-            except Exception:
-                logger.error('Error in loop', exc_info=True)
-                await self.set_agent_state_to(AgentState.ERROR)
-                await self.add_error_to_history(
-                    'Oops! Something went wrong while completing your task. You can check the logs for more info.'
-                )
+                await self._step()
+            except asyncio.CancelledError:
+                logger.info('AgentController task was cancelled')
                 break
-
-            if self._is_stuck():
-                logger.info('Loop detected, stopping task')
-                await self.set_agent_state_to(AgentState.ERROR)
-                await self.add_error_to_history(
-                    'I got stuck into a loop, the task has stopped.'
+            except Exception as e:
+                logger.error(f'Error while running the agent: {e}')
+                await self.report_error(
+                    'There was an unexpected error while running the agent'
                 )
+                await self.set_agent_state_to(AgentState.ERROR)
                 break
-            await asyncio.sleep(
-                0.001
-            )  # Give back control for a tick, so other async stuff can run
-        final_state = self.get_agent_state()
-        if final_state == AgentState.RUNNING:
-            await self.set_agent_state_to(AgentState.PAUSED)
+
+            await asyncio.sleep(1)
```

Comment:
This is sleeping much longer than before, is that intentional?

Reply:
The architecture here is very different now.

Before, we called `step()` inside a `for` loop--we'd get an action, synchronously turn it into an observation, then loop again. We'd `sleep` as a tiny hack to make sure we didn't hog the event loop.

Now, the observations are generated asynchronously. The agent_controller tosses an action out into the EventStream, and eventually gets one back.

So now the `step()` function is called once per second, and most of the time it just says "I'm still waiting for the result of my last action".

But now that I'm writing it out--that might not even take a full second. So we can probably drop this to something lower!
```suggestion
            await asyncio.sleep(.1)
```

Reply:
Sounds good!